doc/codegen/modular_codegen_design.md

```diff
@@ -0,0 +1,561 @@
+
+1. **Metadata Collection**
+   - Free symbol analysis (`free_symbols()`)
+   - Symbol and constant resolution
+   - Argument list construction
+   - Shared transient identification
+
+2. **Allocation Analysis**
+   - Allocation lifetime determination (`determine_allocation_lifetime()`)
+   - Allocation scope determination
+   - Memory layout analysis
+   - Persistent/global memory handling
+
+3. **Code Structure Creation**
+   - State struct generation
+   - Function signature creation
+   - Header/footer generation
+   - Constants definition
+
+4. **Traversal and Generation**
+   - Recursive SDFG traversal
+   - State code generation dispatch
+   - Target-specific code routing
+   - Memory allocation/deallocation code
+
+## Proposed Pass Decomposition
+
+### Phase 1: Analysis Passes
+
+#### 1. **TypeInferencePass**
+- **Purpose**: Infer connector types and set default storage/schedule types
+- **Input**: Raw SDFG
+- **Output**: SDFG with inferred types, pipeline_results["type_info"]
+- **Current Location**: `infer_types.py` functions
+
+#### 2. **LibraryExpansionPass**
+- **Purpose**: Expand all library nodes that haven't been expanded
+- **Input**: Type-inferred SDFG
+- **Output**: SDFG with expanded library nodes
+- **Current Location**: `sdfg.expand_library_nodes()`
+
+#### 3. **MetadataCollectionPass**
+- **Purpose**: Collect free symbols, argument lists, constants, shared transients
+- **Input**: Expanded SDFG
+- **Output**: pipeline_results["metadata"] = {symbols, arglist, constants, shared_transients}
+- **Current Location**: `DaCeCodeGenerator.__init__()`
+
+#### 4. **AllocationAnalysisPass**
+- **Purpose**: Determine allocation lifetimes and scopes for all data containers
+- **Input**: SDFG with metadata
+- **Output**: pipeline_results["allocation_info"] = {to_allocate, where_allocated}
+- **Current Location**: `DaCeCodeGenerator.determine_allocation_lifetime()`
+
+#### 5. **ControlFlowAnalysisPass**
+- **Purpose**: Extract structured control flow from state machines
+- **Input**: SDFG
+- **Output**: pipeline_results["control_flow_tree"]
+- **Current Location**: `control_flow.py` functions
+
+#### 6. **TargetAnalysisPass**
+- **Purpose**: Identify required code generation targets and dispatch routes
+- **Input**: SDFG with metadata
+- **Output**: pipeline_results["targets"] = {required_targets, dispatch_info}
+- **Current Location**: `_get_codegen_targets()`
+
+### Phase 2: Transformation Passes
+
+#### 7. **CopyToMapPass**
+- **Purpose**: Convert complex memory copies to Map nodes where needed
+- **Input**: SDFG with targets identified
+- **Output**: SDFG with transformed copies
+- **Current Location**: `cuda.py` preprocessing, various target preprocessors
+- **Applies To**: GPU strided copies, FPGA transfers
+
+#### 8. **StreamAssignmentPass** (GPU-specific)
+- **Purpose**: Assign CUDA streams for concurrent execution
+- **Input**: GPU-targeted SDFG
+- **Output**: SDFG with stream assignments, pipeline_results["stream_info"]
```

Comment:
Should we consider adding a `Node.metadata` field instead, or is it better to contain all stream information in a replicated dictionary? Please consider applying this logic to similar passes that might use node- or edge-based metadata.